pkg/runtime/grpc/grpc-runtime.go

```diff
@@ -0,0 +1,440 @@
+			Description: "Comma-separated list of nodes to run the gadget on",
+		},
+	}
+}
+
+func (r *Runtime) GlobalParamDescs() params.ParamDescs {
+	return params.ParamDescs{
+		{
+			Key:            ParamConnectionMethod,
+			Description:    "Method that should be used to connect to the Inspektor Gadget nodes",
+			PossibleValues: []string{"kubeapi-server-per-node", "grpc-direct"},
+			DefaultValue:   "kubeapi-server-per-node",
+		},
+	}
+}
+
+type gadgetPod struct {
+	name string
+	node string
+}
+
+func (r *Runtime) getGadgetPods(ctx context.Context, nodes []string) ([]gadgetPod, error) {
+	config, err := utils.KubernetesConfigFlags.ToRESTConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to creating RESTConfig: %w", err)
+	}
+
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set up trace client: %w", err)
+	}
+
+	opts := metav1.ListOptions{LabelSelector: "k8s-app=gadget"}
+	pods, err := client.CoreV1().Pods("gadget").List(ctx, opts)
+	if err != nil {
+		return nil, fmt.Errorf("getting pods: %w", err)
+	}
+
+	if len(pods.Items) == 0 {
+		return nil, fmt.Errorf("no gadget pods found. Is Inspektor Gadget deployed?")
+	}
+
+	if len(nodes) == 0 {
+		res := make([]gadgetPod, 0, len(pods.Items))
+
+		for _, pod := range pods.Items {
+			res = append(res, gadgetPod{name: pod.Name, node: pod.Spec.NodeName})
+		}
+
+		return res, nil
+	}
+
+	// Remove possible duplicates in nodes
+	resMap := map[string]string{}
+	for _, node := range nodes {
+		resMap[node] = ""
+	}
+
+	// Filter nodes
+	for node := range resMap {
+		for _, pod := range pods.Items {
+			if node == pod.Spec.NodeName {
+				resMap[node] = pod.Name
+				break
+			}
+		}
+	}
+
+	res := make([]gadgetPod, 0, len(nodes))
+	for node, pod := range resMap {
+		if pod == "" {
+			log.Warnf("node %q not found or without gadget pod", node)
+			continue
+		}
+		res = append(res, gadgetPod{name: pod, node: node})
+	}
+
+	return res, nil
+}
+
+func (r *Runtime) RunGadget(gadgetCtx runtime.GadgetContext) (map[string][]byte, error) {
+	// Get nodes to run on
+	nodes := gadgetCtx.RuntimeParams().Get(ParamNode).AsStringSlice()
+	pods, err := r.getGadgetPods(gadgetCtx.Context(), nodes)
+	if err != nil {
+		return nil, fmt.Errorf("get gadget pods: %w", err)
+	}
+	if len(pods) == 0 {
+		return nil, fmt.Errorf("no nodes found to run on")
+	}
+
+	if gadgetCtx.GadgetDesc().Type() == gadgets.TypeTraceIntervals {
+		gadgetCtx.Parser().EnableSnapshots(
+			gadgetCtx.Context(),
+			time.Duration(gadgetCtx.GadgetParams().Get(gadgets.ParamInterval).AsInt32())*time.Second,
+			2,
+		)
+	}
+
+	results := make(map[string][]byte)
+	var resultsLock sync.Mutex
+
+	wg := sync.WaitGroup{}
+	for _, pod := range pods {
+		wg.Add(1)
+		go func(pod gadgetPod) {
+			gadgetCtx.Logger().Debugf("running gadget on node %q", pod.node)
+			res, err := r.runGadget(gadgetCtx, pod)
+			if err != nil {
+				gadgetCtx.Logger().Errorf("node %q: %w", pod.node, err)
```

Comment:
> I'm wondering if RunGadget should return an error when it fails on a node. (IMO it should, we can print some of the results but the overall status should be an error because the trace didn't run on all nodes)
>
I agree.

> I don't think we should include the error for each node in "Result", we are already returning an error hence we can use it to report all errors.
> 
If we do the above, then yes, it makes sense.

Reply:
> I don't think we should include the error for each node in "Result", we are already returning an error hence we can use it to report all errors. Perhaps https://pkg.go.dev/golang.org/x/sync/errgroup or https://pkg.go.dev/github.com/uber-go/multierr are useful for it.

I think it still make sense to have that information (which node returned what error) available a little longer (and outside of RunGadget()) - if we ever choose to change how those errors are displayed (using a separate pane, status line, or using another frontend), it's easier to format it properly this way. If we just combine it into a single error, that information is more or less lost.

Regarding returning an error or not in the first place:
We discussed even stopping the whole gadget run when one node returns an error - in order to not lose that information and properly display that _something_ went wrong. However, I also see the use case, where the user is aware of an unresponsive node (or other reasons something might fail within a gadget run) and just wants to continue debugging / get the stream of all other nodes. Maybe we could handle both use cases using a flag like `--ignore-errors`/`--ignore-node-errors` or `--onerror=fail`, which would change that behavior accoring to the user.
WDYT?

Reply:
> I think it still make sense to have that information (which node returned what error) available a little longer (and outside of RunGadget()) - if we ever choose to change how those errors are displayed (using a separate pane, status line, or using another frontend), it's easier to format it properly this way. If we just combine it into a single error, that information is more or less lost.
> 

Your point makes sense to me. TBH I'm wondering if `RunGadget()` should return `map[string][]byte`, it seems off that almost all gadgets send the information using the event handler but few of them return the result through this interface. IMO we should uniform that. However, I don't want to block this PR on that, if you think Jose's proposal on https://github.com/inspektor-gadget/inspektor-gadget/pull/1396#discussion_r1143983760 makes sense, that's fine for me. We can revisit this later on. 

> Regarding returning an error or not in the first place: We discussed even stopping the whole gadget run when one node returns an error - in order to not lose that information and properly display that _something_ went wrong. However, I also see the use case, where the user is aware of an unresponsive node (or other reasons something might fail within a gadget run) and just wants to continue debugging / get the stream of all other nodes. Maybe we could handle both use cases using a flag like `--ignore-errors`/`--ignore-node-errors` or `--onerror=fail`, which would change that behavior accoring to the user. WDYT?

I don't think we need a new flag for it. I think we should print log messages when something happens on a node, but not a hard error or stopping the gadget, as you mentioned, it's better to provide some information than just erroring out. I know for toppers this is more difficult as we're cleaning the screen and that message is lost but, as you mentioned somewhere else, that should be fixed by changing the way we output data.

Reply:
> > I think it still make sense to have that information (which node returned what error) available a little longer (and outside of RunGadget()) - if we ever choose to change how those errors are displayed (using a separate pane, status line, or using another frontend), it's easier to format it properly this way. If we just combine it into a single error, that information is more or less lost.
> 
> Your point makes sense to me. TBH I'm wondering if `RunGadget()` should return `map[string][]byte`, it seems off that almost all gadgets send the information using the event handler but few of them return the result through this interface. IMO we should uniform that. However, I don't want to block this PR on that, if you think Jose's proposal on [#1396 (comment)](https://github.com/inspektor-gadget/inspektor-gadget/pull/1396#discussion_r1143983760) makes sense, that's fine for me. We can revisit this later on.

Great, I will go ahead with @blanquicet's proposal for now, then. I totally agree that we should uniform the output handling as much as possible (after refactoring) - but we should also keep the ability to handle arbitrary data (and convert it using transformers) - not just for the histograms in profile/block-io (which right now come as return value) but also for tcpdump for example, where we could use the raw pcapng data.

> 
> > Regarding returning an error or not in the first place: We discussed even stopping the whole gadget run when one node returns an error - in order to not lose that information and properly display that _something_ went wrong. However, I also see the use case, where the user is aware of an unresponsive node (or other reasons something might fail within a gadget run) and just wants to continue debugging / get the stream of all other nodes. Maybe we could handle both use cases using a flag like `--ignore-errors`/`--ignore-node-errors` or `--onerror=fail`, which would change that behavior accoring to the user. WDYT?
> 
> I don't think we need a new flag for it. I think we should print log messages when something happens on a node, but not a hard error or stopping the gadget, as you mentioned, it's better to provide some information than just erroring out. I know for toppers this is more difficult as we're cleaning the screen and that message is lost but, as you mentioned somewhere else, that should be fixed by changing the way we output data.

👍